rviz_common/src/rviz_common/depth_cloud_mld.cpp

```diff
@@ -0,0 +1,601 @@
+  uint32_t roi_height =
+    camera_info_msg->roi.height > 0 ? camera_info_msg->roi.height : camera_info_msg->height;
+
+  uint32_t expected_width = roi_width / binning_x;
+  uint32_t expected_height = roi_height / binning_y;
+
+  if (expected_width != depth_msg->width || expected_height != depth_msg->height) {
+    std::ostringstream s;
+    s << "Depth image size and camera info don't match: ";
+    s << depth_msg->width << " x " << depth_msg->height;
+    s << " vs " << expected_width << " x " << expected_height;
+    s << "(binning: " << binning_x << " x " << binning_y;
+    s << ", ROI size: " << roi_width << " x " << roi_height << ")";
+    throw(rviz_common::MultiLayerDepthException(s.str()));
+  }
+
+  uint32_t width = depth_msg->width;
+  uint32_t height = depth_msg->height;
+
+  uint32_t size = width * height;
+
+  if (size != static_cast<uint32_t>(shadow_depth_.size())) {
+    // Allocate memory for shadow processing
+    shadow_depth_.resize(size, 0.0f);
+    shadow_timestamp_.resize(size, 0.0);
+    shadow_buffer_.resize(size * POINT_STEP, 0);
+
+    // Precompute 3D projection matrix
+    //
+    // The following computation of center_x,y and fx,fy duplicates
+    // code in the image_geometry package, but this avoids a dependency
+    // on OpenCV, which simplifies releasing rviz.
+
+    // Combine unit conversion (if necessary) with scaling by focal length for computing (X,Y)
+    double scale_x = camera_info_msg->binning_x > 1 ? (1.0 / camera_info_msg->binning_x) : 1.0;
+    double scale_y = camera_info_msg->binning_y > 1 ? (1.0 / camera_info_msg->binning_y) : 1.0;
+
+    // Use correct principal point from calibration
+    float center_x =
+      static_cast<float>((camera_info_msg->p[2] - camera_info_msg->roi.x_offset) * scale_x);
+    float center_y =
+      static_cast<float>((camera_info_msg->p[6] - camera_info_msg->roi.y_offset) * scale_y);
+
+    double fx = camera_info_msg->p[0] * scale_x;
+    double fy = camera_info_msg->p[5] * scale_y;
+
+    float constant_x = static_cast<float>(1.0f / fx);
+    float constant_y = static_cast<float>(1.0f / fy);
+
+    projection_map_x_.resize(width);
+    projection_map_y_.resize(height);
+    std::vector<float>::iterator projX = projection_map_x_.begin();
+    std::vector<float>::iterator projY = projection_map_y_.begin();
+
+    // precompute 3D projection matrix
+    for (uint32_t v = 0; v < height; ++v, ++projY) {
+      *projY = (v - center_y) * constant_y;
+    }
+
+    for (uint32_t u = 0; u < width; ++u, ++projX) {
+      *projX = (u - center_x) * constant_x;
+    }
+
+    // reset shadow vectors
+    reset();
+  }
+}
+
+
+template<typename T>
+sensor_msgs::msg::PointCloud2::SharedPtr
+MultiLayerDepth::generatePointCloudSL(
+  const sensor_msgs::msg::Image::ConstSharedPtr & depth_msg,
+  std::vector<uint32_t> & rgba_color_raw)
+{
+  int width = depth_msg->width;
+  int height = depth_msg->height;
```

Comment:
```suggestion
  uint32_t width = depth_msg->width;
  uint32_t height = depth_msg->height;
```

(actually, you could just drop these and directly use `depth_msg->width` and `depth_msg->height` in the single place they are used below)